File: Python/ByTier/A/2037_A_Twice.py
# Problem: CF 2037 A - Twice
# https://codeforces.com/contest/2037/problem/A

# ================================================================================
# Problem: Maximum Score from Pairing Equal Elements
#
# Algorithm/Techniques:
#   - Frequency counting using set and count()
#   - Greedy approach: For each unique element, we can form pairs equal to floor(count/2)
#
# Time Complexity: O(n^2) per test case
#   - Loop over lines: O(n)
#   - For each line, split and count occurrences: O(n^2)
# Space Complexity: O(n)
#   - Storing the input array and set of unique elements
#
# ================================================================================

for s in [*open(0)][2::2]:  # Read every second line starting from index 2 (skip n and empty lines)
    a = s.split()           # Split the string into list of integers
    print(sum(a.count(x) // 2 for x in {*a}))  # Count pairs for each unique element


# https://github.com/VaHiX/codeForces/

File: CSharp/ByTier/A/2037_A_Twice.cs
// Problem: CF 2037 A - Twice
// https://codeforces.com/contest/2037/problem/A

/*
 * Problem: Maximum Score by Pairing Equal Elements
 *
 * Description:
 * Given an array of integers, we want to find the maximum number of pairs (i,j) such that a[i] == a[j]
 * and i < j. Each element can only be used once in a pair.
 *
 * Algorithm:
 * - Count the frequency of each element in the array.
 * - For each element with frequency f, we can form f / 2 pairs.
 * - Sum all such pairs to get the result.
 *
 * Time Complexity: O(n), where n is the length of the array.
 * Space Complexity: O(n), due to the frequency counter array.
 */

using System;
public class hello {
  static void Main() {
    var T = int.Parse(Console.ReadLine().Trim()); // Read number of test cases
    while (T-- > 0) { // Process each test case
      var n = int.Parse(Console.ReadLine().Trim()); // Read size of array
      string[] line = Console.ReadLine().Trim().Split(' '); // Read array elements
      var a = Array.ConvertAll(line, int.Parse); // Convert string array to int array
      getAns(n, a); // Compute and print answer for current test case
    }
  }
  
  static void getAns(int n, int[] a) {
    var c = new int[n + 1]; // Frequency counter array for values 1 to n
    foreach (var x in a)
      c[x]++; // Count frequency of each element
    
    var ans = 0;
    foreach (var x in c)
      ans += x / 2; // Add number of pairs possible for each element
    
    Console.WriteLine(ans);
  }
}


// https://github.com/VaHiX/codeForces/